solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -719,6 +719,7 @@ public void load() {
     createHandler(ZK_PATH, ZookeeperInfoHandler.class.getName(), ZookeeperInfoHandler.class);
     createHandler(ZK_STATUS_PATH, ZookeeperStatusHandler.class.getName(), ZookeeperStatusHandler.class);
     collectionsHandler = createHandler(COLLECTIONS_HANDLER_PATH, cfg.getCollectionsHandlerClass(), CollectionsHandler.class);
+    healthCheckHandler = loader.newInstance(cfg.getHealthCheckHandlerClass(), HealthCheckHandler.class, null, new Class<?>[]{CoreContainer.class}, new Object[]{this});
```

Comment:
If the intent here to use loader.newInstance is for pluggable implementations, then this may not work with packages. Hence, this should be avoided.

Reply:
> why are we calling loader.newInstance directly instead of createHandler?

I did this because `createHandler` also registers the handler to a path and initializes metrics. HealthCheckHandler currently is ran inside InfoHandler, so that's where the path is registered (and I guess the metrics are merged there). I didn't want to move to another path as part of this PR, don't know if that's needed. True that it looks a bit weird, but otherwise requires a bunch of changes to support compatibility.
> If the intent here to use loader.newInstance is for pluggable implementations, then this may not work with packages. Hence, this should be avoided.

I can handle that? I don't see any of the other top level handlers load from packages other than the built in (which is what `loader.newInstance`. What do you suggest?

Reply:
@chatman, looking a bit at the Packages code. Is the idea that we need to use `PackageListeningClassLoader`? That's not currently used in CoreContainer, and I only see it for loading schema plugins. Maybe I'm missing something. If none of the existing top level handlers can be loaded from packages, then we probably want to tackle that in a different Jira issue?

Reply:
> ... I did this because `createHandler` also registers the handler to a path and initializes metrics. HealthCheckHandler currently is ran inside InfoHandler, so that's where the path is registered (and I guess the metrics are merged there). ...

I wonder if it would be helpful to have that as a comment in the code for future reference, the bit about not wishing to register to a path and initialise metrics.

As far as I can tell `InfoHandler` does not do any metric initialisation, so the health check handler would then be consistent with the other handlers whose path is registered in `InfoHandler`. Or maybe I'm missing how the metrics initialisation happens indirectly?

Reply:
> As far as I can tell InfoHandler does not do any metric initialisation,

From what I can see, `InfoHandler` extends `SolrMetricProducer`, so a call to `CoreContainer.createHandler`  will initialize it (seems to be getting everything from`RequestHandlerBase`

bq. I wonder if it would be helpful to have that as a comment in the code for future reference, the bit about not wishing to register to a path and initialise metrics.
Will do